pkg/image/api/validation/validation.go

```diff
@@ -59,9 +60,74 @@ func validateImage(image *api.Image, fldPath *field.Path) field.ErrorList {
+		allErrs = append(allErrs, field.Required(fldPath.Child("type"), ""))
+	}
+	if len(signature.Content) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("content"), ""))
+	}
+
+	var (
+		trustedCondition, forImageCondition *api.SignatureCondition
+		forImageConditionIndex              = -1
+	)
+	for i := range signature.Conditions {
+		cond := &signature.Conditions[i]
+		if cond.Type == api.SignatureTrusted && (trustedCondition == nil || !cond.LastProbeTime.Before(trustedCondition.LastProbeTime)) {
+			trustedCondition = cond
+		} else if cond.Type == api.SignatureForImage && forImageCondition == nil || !cond.LastProbeTime.Before(forImageCondition.LastProbeTime) {
+			forImageCondition = cond
+			forImageConditionIndex = i
+		}
+	}
+
+	if trustedCondition != nil && forImageCondition == nil {
+		msg := fmt.Sprintf("missing %q condition type", api.SignatureForImage)
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), signature.Conditions, msg))
+	} else if forImageCondition != nil && trustedCondition == nil {
+		msg := fmt.Sprintf("missing %q condition type", api.SignatureTrusted)
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), signature.Conditions, msg))
+	} else if trustedCondition != nil && forImageCondition != nil {
+		if ((trustedCondition.Status == kapi.ConditionUnknown) && (forImageCondition.Status != kapi.ConditionUnknown)) || ((trustedCondition.Status == kapi.ConditionTrue) && (forImageCondition.Status != kapi.ConditionTrue)) {
+			msg := fmt.Sprintf("invalid status of %q condition for %s=%s", api.SignatureForImage, api.SignatureTrusted, trustedCondition.Status)
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions").Index(forImageConditionIndex).Child("status"), forImageCondition.Status, msg))
+		}
+	}
+
+	if trustedCondition == nil || trustedCondition.Status == kapi.ConditionUnknown {
+		if len(signature.ImageIdentity) != 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("imageIdentity"), signature.ImageIdentity, "must be unset for unknown signature state"))
+		}
+		if len(signature.SignedClaims) != 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("signedClaims"), signature.SignedClaims, "must be unset for unknown signature state"))
+		}
+		if !reflect.DeepEqual(&signature.IssuedBy, &api.SignatureIssuer{}) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("issuedBy"), signature.IssuedBy, "must be unset for unknown signature state"))
+		}
+		if !reflect.DeepEqual(&signature.IssuedTo, &api.SignatureSubject{}) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("issuedTo"), signature.IssuedTo, "must be unset for unknown signature state"))
+		}
+	}
+
+	if trustedCondition != nil && trustedCondition.Status == kapi.ConditionTrue {
+		if len(signature.ImageIdentity) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("imageIdentity"), signature.ImageIdentity, "ImageIdentity must be set for known signature state"))
```

Comment:
Removed.
On 06/06/2016 05:31 PM, David Eads wrote:
>
> In pkg/image/api/validation/validation.go <https://github.com/openshift/origin/pull/8371#discussion_r65911949>:
>
> > +			allErrs = append(allErrs, field.Invalid(fldPath.Child("imageIdentity"), signature.ImageIdentity, "must be unset for unknown signature state"))
> > +		}
> > +		if len(signature.SignedClaims) != 0 {
> > +			allErrs = append(allErrs, field.Invalid(fldPath.Child("signedClaims"), signature.SignedClaims, "must be unset for unknown signature state"))
> > +		}
> > +		if !reflect.DeepEqual(&signature.IssuedBy, &api.SignatureIssuer{}) {
> > +			allErrs = append(allErrs, field.Invalid(fldPath.Child("issuedBy"), signature.IssuedBy, "must be unset for unknown signature state"))
> > +		}
> > +		if !reflect.DeepEqual(&signature.IssuedTo, &api.SignatureSubject{}) {
> > +			allErrs = append(allErrs, field.Invalid(fldPath.Child("issuedTo"), signature.IssuedTo, "must be unset for unknown signature state"))
> > +		}
> > +	}
> > +
> > +	if trustedCondition != nil && trustedCondition.Status == kapi.ConditionTrue {
> > +		if len(signature.ImageIdentity) == 0 {
> > +			allErrs = append(allErrs, field.Invalid(fldPath.Child("imageIdentity"), signature.ImageIdentity, "ImageIdentity must be set for known signature state"))
>
> I think this is too strict.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openshift/origin/pull/8371/files/4371c2172d2421e632faa67cf851acce66753c04#r65911949>, or mute the 
> thread <https://github.com/notifications/unsubscribe/ABaQk6UX454J7z1Cx_Jd4C4O2CyINlkvks5qJD1SgaJpZM4IBAFQ>.
>